data/blog/battery-life.mdx

```diff
@@ -0,0 +1,86 @@
+---
+title: 'Battery life instrumentation'
+date: '2023-04-25'
+tags:
+  - 'oss'
+  - 'mobile'
+authors: ['mihai']
+images: ['/images/battery-life/social.png']
+summary: 'Instrumentation that we’re adding to the Tailscale clients to have a scalable approach to battery life investigations.'
+---
+
+Battery consumption is one of the top areas where we’ve head that Tailscale needs to improve, via both [GitHub](https://github.com/tailscale/tailscale/issues/3363) and [social media](https://twitter.com/search?q=tailscale%20battery&src=typed_query&f=live). There’s always been [work on battery-related issues](https://github.com/tailscale/tailscale/issues?q=label%3Abattery+is%3Aclosed), but it’s not something that can ever be declared “done”:
+
+- Tailscale’s feature set evolves — using the [NextDNS integration](https://tailscale.com/blog/nextdns/) also means that Tailscale is now responsible for all DNS requests issued by a device.
+- The operating system changes, and Apple or Google may intentionally or [unintentionally](https://github.com/tailscale/tailscale/issues/4646#issuecomment-1257251332) change how the application operates and when it’s allowed to run, and how much power it consumes.
+- New patterns of user behavior may arise, for example there may new reasons to [use an exit node](https://www.youtube.com/watch?v=9CunwUs08og) more frequently.
+
+One of the challenges in this work is that it’s hard to get aggregated and actionable data. It’s great that recent versions of iOS and Xcode will collect and aggregate [metrics about battery use](https://developer.apple.com/documentation/xcode/analyzing-your-app-s-battery-use), but the numbers are delayed and very coarse-grained. In the graph below, it’s great that Tailscale 1.36 showed a marked improvement, but it’s not clear why (perhaps it was due to the [interface binding changes](https://tailscale.dev/blog/darwin-spelunking), or perhaps it was something else). If the direction had been reversed, we would be hard-pressed to know where to begin investigating.
```

Comment:
```suggestion
One of the challenges in this work is that it’s hard to get aggregated and actionable data. It’s great that recent versions of iOS and Xcode will collect and aggregate [metrics about battery use](https://developer.apple.com/documentation/xcode/analyzing-your-app-s-battery-use), but the numbers are delayed and very coarse-grained. In the graph below, Tailscale v1.36 showed a marked improvement compared to v1.34, but it’s not clear why (perhaps it was due to the [interface binding changes](https://tailscale.dev/blog/darwin-spelunking), or perhaps it was something else). If the direction had been reversed, we would be hard-pressed to know where to begin investigating.
```
Did we do something specific in 1.36? It sounds like we are more pleasantly surprised than did something to fix it.

Reply:
We did do something (the interface binding changes), but we were not confident that they would show up in the graphs (and have no way to be sure, since they're so coarse-grained). Tweaked the text a bit.